### ISSUE TITLE
CodeGen: use await in generated IGrainMethodInvoker.Invoke methods & avoid .Box()

### ISSUE THREAD

**Issue by ReubenBond** [dotnet/orleans#1074]
Currently, generated `IGrainMethodInvoker` implementations look something like this:
```C#
publicTask<Object> Invoke(IAddressable grain, Int32 interfaceId, Int32 methodId, Object[] arguments)
{
  try
  {
    if (grain == null)
      throw new ArgumentNullException("grain");
    switch (interfaceId)
    {
      case -1135060418:
        switch (methodId)
        {
          case -355259193:
            return ((IReminderTableGrain)grain).ReadRows((GrainReference)arguments[0]).Box();
          case 120951690:
            return ((IReminderTableGrain)grain).TestOnlyClearTable().Box();
          default:
            throw new NotImplementedException("interfaceId=" + -1135060418 + ",methodId=" + methodId);
        }

      default:
        throw new NotImplementedException("interfaceId=" + interfaceId);
    }
  }
  catch (Exception exception)
  {
    return Async.TaskUtility.Faulted(exception);
  }
}
```
The call to `.Box()` involves a trip through the task scheduler, which increases latency for calls and involves an additional allocation and a few dereferences.
In order to improve this situation, I propose that we emit code more similar to the following, which can help especially in cases where grains return already-completed tasks.
```C#
public async Task<Object> Invoke(IAddressable grain, Int32 interfaceId, Int32 methodId, Object[] arguments)
{
  if (grain == null)
    throw new ArgumentNullException("grain");
  switch (interfaceId)
  {
    case -1135060418:
      switch (methodId)
      {
        case -355259193:
          return await ((IReminderTableGrain)grain).ReadRows((GrainReference)arguments[0]);
        case 120951690:
          return await ((IReminderTableGrain)grain).TestOnlyClearTable();
        default:
          throw new NotImplementedException("interfaceId=" + -1135060418 + ",methodId=" + methodId);
      }

    default:
      throw new NotImplementedException("interfaceId=" + interfaceId);
  }
}
```

Note that we remove the outer `try`/`catch` and the call to `.Box()`.
Any concerns?

**Comment by ReubenBond** [comment]
By the way, if anyone else wants experience working with code generation, I am happy to assist :smile: I promise it will be fun! Otherwise I will eventually do this... when I find time.

**Comment by dVakulen** [comment]
I'd like to gain that expirience.

**Comment by ReubenBond** [comment]
Great, @dVakulen!
I find that [Roslyn Quoter](roslynquoter.azurewebsites.net) and the [Syntax Visualizer](https://roslyn.codeplex.com/wikipage?title=Syntax%20Visualizer) in Visual Studio 2015 are very handy for working out how the AST should look.

The bulk of this PR is in [GrainMethodInvokerGenerator.cs](https://github.com/dotnet/orleans/blob/52975575a1a10ba95b323b15f2c78b17e7e441c2/src/OrleansCodeGenerator/GrainMethodInvokerGenerator.cs), particularly in `GenerateInvokeMethod` where you'll want to remove the `try`/`catch`, `.Box()` and add the `async` & `await`s.

We may as well completely nuke the `Box()` method - I don't like that it exists in the public scope.

Thank you very much & feel free to DM me questions or ask them here!

**Comment by dVakulen** [comment]
@ReubenBond Thanks for the info! One question: do we really need to await on return? The ```Box()``` method is only adds continuation, and not waits for the original task to complete.

**Comment by sergeybykov** [comment]
`async/await` adds allocation of a state machine. So the perf can be worse in this case, unless I missed something. The general recommendation from the TPL folks has always been to only use `await` when necessary. I don't know if that was improved in .NET 4.6 or not.

**Comment by ReubenBond** [comment]
@sergeybykov we are allocating anyway, by virtue of the continuation in `Box`. `async`/`await` means that we do not need to take another trip through the Task scheduler, which adds latency.

**Comment by sergeybykov** [comment]
We can tests and see if there's any noticeable gain/loss. Doesn't `await` in general also incur a trip through the scheduler though?

**Comment by ReubenBond** [comment]
`await` will avoid that trip if the task executes synchronously, which is the case for many tasks in Orleans, since we often return `Task.FromResult(this.State.X)`
Here's a quick test for that: https://gist.github.com/ReubenBond/445a7cd47bf403b323f6

**Comment by dVakulen** [comment]
I still don't get why there's need in ```await``` in generated invoke. The returned task is being awaited in the following places: https://github.com/dotnet/orleans/blob/master/src/Orleans/Runtime/OutsideRuntimeClient.cs#L474 and https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Core/InsideGrainClient.cs#L387.

**Comment by ReubenBond** [comment]
See my [comment above](https://github.com/dotnet/orleans/issues/1074#issuecomment-159711873), @dVakulen. Basically it avoids resheduling, which is a source of latency, particularly under load.

**Comment by gabikliot** [comment]
@dVakulen , the CW (ContinueWith) or await are needed to cast `Task<T>` to `Task<object>`. Unfortunately, the only we way to do this Cast is via CW or await.

@ReubenBond , before Box we used CW. Now CW is inside Box, but it is still CW.
The theory was that CW is more efficient than await, since await also does CW + allocates a class to hold state (state machine). I do think the general statement of: "Plain CW is more efficient than await" is true. But the difference may be negligible, so it may not do any perf. harm to move to await.
Just need to measure.

**Comment by ReubenBond** [comment]
`await` can be more efficient than `ContinueWith`. For example, if the antecedent completes synchronously (as is the case for many Orleans methods). `await` will return a cached task if possible (good for our `return null` case for void `Task` methods).

`await` defers allocation until the last possible moment in order to reduce the cost.

`ContinueWith` will always allocate.

**Comment by gabikliot** [comment]
OK, I was not aware of all those `await` optimizations. Thanks for educating me.

**Comment by ReubenBond** [comment]
I [extended the benchmark](https://gist.github.com/ReubenBond/445a7cd47bf403b323f6) to include
1. pushing the await into the Box method (thus not eliminating it)
2. including a Box method which passes `ExecuteSynchronously`

Here are the results:
```
 01836ms await
 01938ms await in Box method
 05656ms Box with ExecuteSynchronously
 21207ms Box (current Orleans code)
```

**Comment by jdom** [comment]
so does it mean that we could fix this by just rewriting Box instead of changing generated code? that'd be a cool quick fix

**Comment by ReubenBond** [comment]
@jdom, yes, we can gain most of the benefit by not putting `await` into generated code.
We should be able to gain even more performance by using a hybrid approach, where the `Box` method checks to see that the task has already synchronously completed, if so, returns without await or CW, and otherwise returns a Box method which awaits.
Here are results from the [updated benchmark](https://gist.github.com/ReubenBond/445a7cd47bf403b323f6):
```
 01334ms await
 01314ms await in Box method
 05106ms Box with ExecuteSynchronously
 00250ms Box with completed task optimization
 18810ms Box(current Orleans code)

Non-synchronously completing tasks:
 25627ms await
 25932ms await in Box method
 28156ms Box with ExecuteSynchronously
 24227ms Box with completed task optimization
 33796ms Box(current Orleans code)
```
"Box with completed task optimization" is the new method.

**Comment by gabikliot** [comment]
Sounds pretty convincing.
Just for completeness, to be 100% sure, can you also add a case with a simple CW, without Box method and without `t.GetAwaiter().GetResult()`: just the way it was before: inside the `Invoke` make a grain call and do `task.ContinueWith(t => t.Result)`.

**Comment by ReubenBond** [comment]
Sure thing! I added "Inline ContinueWith" & [updated the Gist](https://gist.github.com/ReubenBond/445a7cd47bf403b323f6).
As you can see, the ContinueWith version is about the same as the `.Box` version used in Orleans:
```
  1275ms await
  1279ms await in Box method
  4890ms Box with ExecuteSynchronously
   249ms Box with completed task optimization
 19621ms Inlined ContinueWith
 20506ms Box (current Orleans code)

Non-synchronously completing tasks:
 25436ms await
 23345ms await in Box method
 27460ms Box with ExecuteSynchronously
 23683ms Box with completed task optimization
 34588ms Inlined ContinueWith
 33807ms Box (current Orleans code)
```

**Comment by gabikliot** [comment]
:+1:

**Comment by dVakulen** [comment]
```MethodInvoker``` sometimes happens to directly invoke method on original grain instance bypassing call from generated code to https://github.com/dotnet/orleans/blob/master/src/Orleans/Runtime/GrainReference.cs#L321, doesn't it?

**Comment by jdom** [comment]
@ReubenBond this is great! For full complete completeness, would you mind testing all combinations? I mean, do both synchronous and asynchronous versions of all the following return types:
- Task
- Task<int>
- Task<string> or some other reference type
- Task<object>

**Comment by gabikliot** [comment]
@dVakulen , grain reference is a client (caller) stub. A call from the application is always done via the stub (grain reference). It always send a real message. When the message arrives it is dispatched/invoked via `MethodInvoker ` on the activation object.

**Comment by ReubenBond** [comment]
@jdom here you go :)
Note: I added a new optimization to BoxMaySync (the suggested solution) by specializing on Task<object> to return the input directly with no change.
Note 2: Exception handling is slow, so I reduced the iteration for the 'Faulted' and 'Cancelled' tests by a factor of 100 to make them palatable.
[Gist updated](https://gist.github.com/ReubenBond/445a7cd47bf403b323f6)
```
Variant: await
    1352ms (method: completed Task)
    1419ms (method: completed Task<int>)
    1440ms (method: completed Task<string>)
    1410ms (method: completed Task<object>)
   24750ms (method: yielded Task)
   25377ms (method: yielded Task<int>)
   25125ms (method: yielded Task<string>)
   25136ms (method: yielded Task<object>)
    7349ms (method: Faulted)
    7861ms (method: Cancelled)

Variant: await in Box
    1298ms (method: completed Task)
    1373ms (method: completed Task<int>)
    1375ms (method: completed Task<string>)
    1381ms (method: completed Task<object>)
   22894ms (method: yielded Task)
   23731ms (method: yielded Task<int>)
   23077ms (method: yielded Task<string>)
   23351ms (method: yielded Task<object>)
    7466ms (method: Faulted)
    8032ms (method: Cancelled)

Variant: Box ExecuteSynchronously
    4894ms (method: completed Task)
    4743ms (method: completed Task<int>)
    4841ms (method: completed Task<string>)
    4845ms (method: completed Task<object>)
   27487ms (method: yielded Task)
   27613ms (method: yielded Task<int>)
   27264ms (method: yielded Task<string>)
   27455ms (method: yielded Task<object>)
    7679ms (method: Faulted)
    8319ms (method: Cancelled)

Variant: Box optimized
     254ms (method: completed Task)
     302ms (method: completed Task<int>)
     303ms (method: completed Task<string>)
     191ms (method: completed Task<object>)
   22274ms (method: yielded Task)
   23020ms (method: yielded Task<int>)
   22489ms (method: yielded Task<string>)
   17138ms (method: yielded Task<object>)
    4230ms (method: Faulted)
    3887ms (method: Cancelled)

Variant: Inlined ContinueWith
   18296ms (method: completed Task)
   18589ms (method: completed Task<int>)
   18212ms (method: completed Task<string>)
   18268ms (method: completed Task<object>)
   32316ms (method: yielded Task)
   33031ms (method: yielded Task<int>)
   32518ms (method: yielded Task<string>)
   32215ms (method: yielded Task<object>)
   11514ms (method: Faulted)
   11512ms (method: Cancelled)

Variant: Box (Current Orleans)
   18421ms (method: completed Task)
   18785ms (method: completed Task<int>)
   18621ms (method: completed Task<string>)
   18166ms (method: completed Task<object>)
   32516ms (method: yielded Task)
   34526ms (method: yielded Task<int>)
   33374ms (method: yielded Task<string>)
   33678ms (method: yielded Task<object>)
   12313ms (method: Faulted)
   11358ms (method: Cancelled)
```

![image](https://cloud.githubusercontent.com/assets/203839/11419561/5699cbe2-947c-11e5-96f7-570e3c3e4bc4.png)

**Comment by gabikliot** [comment]
I think what I actually learnt from your gist code, is on that line:
https://gist.github.com/ReubenBond/445a7cd47bf403b323f6#file-program-cs-L201
you just call the `BoxAwait` and `await` inside, but you don't `await` the `BoxAwait` call.
That means there is no await/CW/state machine at all happening in the sync case.
If instead you were awating the `BoxAwait` in the `BoxMaySync`, then there would be state machine allocated any way, even if the code returns synchronously.
:+1: 

If only Java had this stuff !!

**Comment by dVakulen** [comment]
I see. In ```MethodIvoker``` grain can be either of type ```GrainReference``` or ```Grain```. So I suggesting to consider the following approach: change ```IGrainMethodInvoker.cs```: ```Task<Object> Invoke(IAddressable grain, Int32 interfaceId, Int32 methodId, Object[] arguments)``` to accept ```TaskCompletionSource<object>``` as parameter. For the ```Grain``` calls apply  suggested by @ReubenBond solution, and for ```GrainReference``` ones - make direct call to the ```GrainReference.InvokeMethodAsync``` in which tcs will be initialized in the OrleansTaskExtentions.cs  ```Task<T> ConvertTaskViaTcs<T>(Task<T> task, TaskCompletionSource<object> resolver)```, thus removing the need in additional ```await```:
https://gist.github.com/dVakulen/4e943ff1a656edf2e58a

**Comment by ReubenBond** [comment]
I updated the tests with @yevhen's suggestion of using [BenchmarkDotNet](https://github.com/PerfDotNet/BenchmarkDotNet) and reran. Here are the results.
*TL;DR:* BoxMaySync wins.

**Throughput** in ops/sec (more is better)
* BoxMaySync blows the others away for completed tasks, eg `Task.FromResult(x)` - a common pattern in Orleans)
* BoxMaySync edges the others out in all but one (close) test for uncompleted tasks.

![image](https://cloud.githubusercontent.com/assets/203839/11432774/17fd0966-9509-11e5-9bbb-b34c0973827b.png)

**Latency** in ns/op (less it better)
* BoxMaySync is better, same results as above.
* Std.Dev. expressed as a percentage of latency was similarly low for all tests.
![image](https://cloud.githubusercontent.com/assets/203839/11432842/0b7c7630-950a-11e5-9fa5-1e840e0895d5.png)

Benchmark output:
```
                 Method | Param |       AvrTime |      StdDev |          op/s |
----------------------- |------ |-------------- |------------ |-------------- |
        InvokeAwaitTest |     1 |   118.1399 ns |   1.1671 ns |  8,464,540.47 |
        InvokeAwaitTest |     2 |   122.9525 ns |   2.4482 ns |  8,133,219.77 |
        InvokeAwaitTest |     3 |   118.8884 ns |   1.8040 ns |  8,411,251.83 |
        InvokeAwaitTest |     4 |   120.1652 ns |   2.6175 ns |  8,321,874.51 |
        InvokeAwaitTest |     5 | 5,881.6269 ns | 137.4339 ns |    170,021.28 |
        InvokeAwaitTest |     6 | 5,895.1784 ns | 142.8124 ns |    169,630.15 |
        InvokeAwaitTest |     7 | 5,894.9718 ns | 107.0907 ns |    169,636.37 |
        InvokeAwaitTest |     8 | 5,853.7815 ns |  87.0614 ns |    170,829.76 |
     InvokeBoxAwaitTest |     1 |   118.6376 ns |   3.6930 ns |  8,429,030.78 |
     InvokeBoxAwaitTest |     2 |   122.1116 ns |   3.3929 ns |  8,189,246.32 |
     InvokeBoxAwaitTest |     3 |   124.8825 ns |   2.2639 ns |  8,007,533.05 |
     InvokeBoxAwaitTest |     4 |   121.8984 ns |   2.4253 ns |  8,203,556.10 |
     InvokeBoxAwaitTest |     5 | 5,781.5249 ns |  78.7605 ns |    172,964.81 |
     InvokeBoxAwaitTest |     6 | 5,939.2875 ns | 147.0792 ns |    168,370.41 |
     InvokeBoxAwaitTest |     7 | 5,807.6403 ns | 114.7573 ns |    172,186.98 |
     InvokeBoxAwaitTest |     8 | 5,784.6669 ns |  79.6665 ns |    172,870.86 |
   InvokeBoxMaySyncTest |     1 |    18.1577 ns |   0.7623 ns | 55,072,934.56 |
   InvokeBoxMaySyncTest |     2 |    23.2627 ns |   0.8242 ns | 42,987,408.60 |
   InvokeBoxMaySyncTest |     3 |    23.3452 ns |   0.8339 ns | 42,835,392.01 |
   InvokeBoxMaySyncTest |     4 |    11.7935 ns |   0.2719 ns | 84,793,445.87 |
   InvokeBoxMaySyncTest |     5 | 5,767.6802 ns | 157.4282 ns |    173,381.08 |
   InvokeBoxMaySyncTest |     6 | 5,894.1003 ns | 111.0824 ns |    169,661.18 |
   InvokeBoxMaySyncTest |     7 | 5,911.8043 ns | 104.7147 ns |    169,153.26 |
   InvokeBoxMaySyncTest |     8 | 5,132.7924 ns | 150.7766 ns |    194,831.81 |
      InvokeBoxSyncTest |     1 |   443.8770 ns |   8.4379 ns |  2,252,876.37 |
      InvokeBoxSyncTest |     2 |   452.3214 ns |   3.9123 ns |  2,210,818.05 |
      InvokeBoxSyncTest |     3 |   459.5984 ns |   4.4377 ns |  2,175,812.66 |
      InvokeBoxSyncTest |     4 |   469.4621 ns |  38.4072 ns |  2,130,116.72 |
      InvokeBoxSyncTest |     5 | 5,905.5574 ns |  83.3330 ns |    169,332.07 |
      InvokeBoxSyncTest |     6 | 5,772.0463 ns |  73.2652 ns |    173,249.14 |
      InvokeBoxSyncTest |     7 | 5,867.2747 ns |  80.0128 ns |    170,436.98 |
      InvokeBoxSyncTest |     8 | 5,858.1026 ns |  88.3311 ns |    170,703.75 |
          InvokeBoxTest |     1 | 5,411.5318 ns |  88.6613 ns |    184,790.93 |
          InvokeBoxTest |     2 | 5,316.1122 ns |  75.5348 ns |    188,107.70 |
          InvokeBoxTest |     3 | 5,339.5211 ns |  82.7055 ns |    187,282.72 |
          InvokeBoxTest |     4 | 5,423.9605 ns |  89.3831 ns |    184,367.12 |
          InvokeBoxTest |     5 | 6,996.8031 ns | 112.6622 ns |    142,922.42 |
          InvokeBoxTest |     6 | 7,007.4688 ns |  98.6143 ns |    142,705.01 |
          InvokeBoxTest |     7 | 7,080.5600 ns | 103.9489 ns |    141,231.98 |
          InvokeBoxTest |     8 | 7,085.8069 ns | 110.8695 ns |    141,127.65 |
 InvokeContinueWithTest |     1 | 5,385.4396 ns | 109.0407 ns |    185,685.92 |
 InvokeContinueWithTest |     2 | 5,462.8252 ns |  75.4047 ns |    183,055.47 |
 InvokeContinueWithTest |     3 | 5,398.7773 ns |  57.4820 ns |    185,227.26 |
 InvokeContinueWithTest |     4 | 5,377.6220 ns |  75.8221 ns |    185,955.89 |
 InvokeContinueWithTest |     5 | 7,005.2245 ns | 138.6961 ns |    142,750.61 |
 InvokeContinueWithTest |     6 | 6,921.4157 ns | 118.7271 ns |    144,479.50 |
 InvokeContinueWithTest |     7 | 7,079.0742 ns |  96.8791 ns |    141,262.30 |
 InvokeContinueWithTest |     8 | 7,107.0925 ns | 101.0203 ns |    140,704.51 |
```

**Comment by yevhen** [comment]
:+1:

**Comment by gabikliot** [comment]
Implemented via #1085 by @dVakulen . Thank you.
